### PR TITLE
Rework impressions strategy to support more use cases

### PR DESCRIPTION
**Pull request by hopsoft**
Add new scopes to support campaign/ad rendering and organize related methods.

### CODE REVIEW

File: app/models/concerns/campaigns/clickable.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    def total_clicks_count_cache_key
+      "#{cache_key}/#{TOTAL_CLICKS_COUNT_KEY}"
+    end
+
+    def daily_clicks_count_cache_key(date = nil)
+      "#{cache_key}/#{DAILY_CLICKS_COUNT_KEY}/#{(date || Date.current).to_date.iso8601}"
+    end
+
+    def total_clicks_count
```
Comment (coderberry) [inline app/models/concerns/campaigns/clickable.rb]: We need to increment when a click occurs

File: app/models/concerns/campaigns/clickable.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      Rails.cache.fetch total_clicks_count_cache_key, expires_in: (remaining_operative_days + 7).days do
+        impressions.clicked.count
+      end
+    end
+
+    def daily_clicks_count(date = nil)
```
Comment (coderberry) [inline app/models/concerns/campaigns/clickable.rb]: We need to increment when a click occurs for the day

File: app/models/concerns/campaigns/displayable.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    def total_impressions_count_cache_key
+      "#{cache_key}/#{TOTAL_IMPRESSIONS_COUNT_KEY}"
+    end
+
+    def daily_impressions_count_cache_key(date = nil)
+      "#{cache_key}/#{DAILY_IMPRESSIONS_COUNT_KEY}/#{(date || Date.current).to_date.iso8601}"
+    end
+
+    def total_impressions_count
```
Comment (coderberry) [inline app/models/concerns/campaigns/displayable.rb]: We need to increment when an impression occurs

File: app/models/concerns/campaigns/displayable.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      Rails.cache.fetch total_impressions_count_cache_key, expires_in: (remaining_operative_days + 7).days do
+        impressions.count
+      end
+    end
+
+    def total_impressions_per_mille
+      total_impressions_count / 1_000.to_f
+    end
+
+    def daily_impressions_count(date = nil)
```
Comment (coderberry) [inline app/models/concerns/campaigns/displayable.rb]: We need to increment when an impression occurs for that date

File: app/models/concerns/campaigns/recommendable.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
+module Campaigns
+  module Recommendable
+    extend ActiveSupport::Concern
+
+    def recommended_daily_budget
+      total_remaining_budget / remaining_operative_days
+    end
+
+    def recommended_end_date
+      total_available_impression_count = ((total_remaining_budget.to_f / ecpm.to_f) * 1_000).ceil
```
Comment (coderberry) [inline app/models/concerns/campaigns/recommendable.rb]: Maybe extract this into a `remaining_impression_count` method?

File: app/models/concerns/campaigns/recommendable.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+      days = total_available_impression_count / estimated_max_daily_impression_count
```
Comment (coderberry) [inline app/models/concerns/campaigns/recommendable.rb]: Maybe extract into it's own method as well?

File: app/models/impression.rb [hunk 7]
```diff
@@ -26,45 +27,80 @@ class Impression < ApplicationRecord
...
-  validates :ip, length: {maximum: 255, allow_blank: false}
-  validates :postal_code, length: {maximum: 255}
 
   # callbacks .................................................................
-  after_commit :increment_counters, on: [:create]
+  before_validation :assure_campaign_name, on: [:create]
+  before_validation :assure_property_name, on: [:create]
+  before_validation :set_displayed_at, on: [:create]
+  before_create :assure_partition_table, on: [:create]
+  after_commit :set_property_advertiser, on: [:create]
```
Comment (coderberry) [inline app/models/impression.rb]: Add redis incrementers

File: app/models/impression.rb [hunk 8]
```diff
@@ -26,45 +27,80 @@ class Impression < ApplicationRecord
...
   # additional config (i.e. accepts_nested_attribute_for etc...) ..............
-  range_partition_by :displayed_at_date
 
   # class methods .............................................................
   class << self
   end
 
   # public instance methods ...................................................
 
+  def partition_table_name
```
Comment (coderberry) [inline app/models/impression.rb]: Maybe add note to README about user requirements for deploy

File: app/models/impression.rb [hunk 9]
```diff
@@ -26,45 +27,80 @@ class Impression < ApplicationRecord
...
   # scopes ....................................................................
-  scope :clicked, -> { where.not(clicked_at: nil) }
-  scope :on, ->(date) { where displayed_at_date: date }
-
-  # This scope may look odd but isolating the date range query like this allows us to
-  # optimize for the fastest query that takes advantage of partitioning and indexes
-  scope :between, ->(start_date, end_date) {
-    where displayed_at_date: (start_date..end_date).map(&:iso8601)
-  }
+  scope :clicked, -> { where.not clicked_at: nil }
```
Comment (coderberry) [inline app/models/impression.rb]: Do we need to add an index for this? (clicked_at and clicked_at_date)?

File: app/models/property.rb [hunk 10]
```diff
@@ -54,6 +60,12 @@ class Property < ApplicationRecord
   scope :search_url, ->(value) { value.blank? ? all : search_column(:url, value) }
   scope :search_user, ->(value) { value.blank? ? all : where(user_id: User.publisher.search_name(value)) }
   scope :search_user_id, ->(value) { value.blank? ? all : where(user_id: value) }
+  scope :for_campaign, ->(campaign) {
+    relation = active.with_any_keywords(*campaign.keywords).without_any_keywords(*campaign.negative_keywords)
+    relation = relation.where(prohibit_fallback_campaigns: false) if campaign.fallback?
+    # TODO: omit prohibitited advertisers
```
Comment (coderberry) [inline app/models/property.rb]: Let's discuss this. This needs some planning

File: app/models/property.rb [hunk 11]
```diff
@@ -54,6 +60,12 @@ class Property < ApplicationRecord
   scope :search_url, ->(value) { value.blank? ? all : search_column(:url, value) }
   scope :search_user, ->(value) { value.blank? ? all : where(user_id: User.publisher.search_name(value)) }
   scope :search_user_id, ->(value) { value.blank? ? all : where(user_id: value) }
+  scope :for_campaign, ->(campaign) {
+    relation = active.with_any_keywords(*campaign.keywords).without_any_keywords(*campaign.negative_keywords)
+    relation = relation.where(prohibit_fallback_campaigns: false) if campaign.fallback?
```
Comment (coderberry) [inline app/models/property.rb]: We should discuss fallback campaigns. I'm not sure how it's currently set up

File: app/models/campaign.rb [hunk 12]
```diff
@@ -135,205 +132,62 @@ class << self
...
-    ENUMS::CAMPAIGN_STATUSES.archived? status
+  def property_ids_with_impressions(date = nil)
+    return Impression.on(date).where(campaign: self).select(:property_id).distinct.pluck(:property_id) if date
+    impressions.select(:property_id).distinct.pluck(:property_id)
   end
 
-  def available_on?(date)
-    date.to_date.between? start_date, end_date
+  def property_ids_with_payable_impressions(date = nil)
+    return Impression.payable.on(date).where(campaign: self).select(:property_id).distinct.pluck(:property_id) if date
```
Comment (coderberry) [inline app/models/campaign.rb]: maybe `impressions.payable.on(date).select(...`

File: app/models/property.rb [hunk 13]
```diff
@@ -109,6 +121,22 @@ def pretty_url
     url.gsub(/^https?:\/\//, "").gsub("www.", "").split("/").first
   end
 
+  def matching_campaigns
+    Campaign.for_property self
+  end
+
+  def advertisers
+    User.where(id: property_advertisers.select(:advertiser_id)) unless date
```
Comment (coderberry) [inline app/models/property.rb]: Return `User.none` if date is not present?

File: app/models/property_advertiser.rb [hunk 14]
```diff
@@ -1,33 +1,38 @@
...
-  belongs_to :record, polymorphic: true
+  belongs_to :advertiser, class_name: "User", foreign_key: "advertiser_id"
+  belongs_to :property
 
   # validations ...............................................................
   # callbacks .................................................................
-
   # scopes ....................................................................
-  scope :segmented_by, ->(value) { where segment: value }
+  scope :on, ->(date) { where displayed_at_date: date.to_date }
```
Comment (coderberry) [inline app/models/property_advertiser.rb]: Remove date from this model

File: app/models/user.rb [hunk 15]
```diff
@@ -64,7 +64,9 @@ class User < ApplicationRecord
   has_many :assets
   has_many :campaigns
   has_many :creatives
+  has_many :impressions, foreign_key: "advertiser_id"
```
Comment (coderberry) [inline app/models/user.rb]: `impressions_as_advertiser` and `impressions_as_publisher` ?

File: app/models/user.rb [hunk 16]
```diff
@@ -64,7 +64,9 @@ class User < ApplicationRecord
...
   has_many :properties
+  has_many :property_advertisers, class_name: "PropertyAdvertiser", foreign_key: "advertiser_id"
```
Comment (coderberry) [inline app/models/user.rb]: Maybe rename b/c it's confusing?

File: db/migrate/20181017152837_shotgun.rb [hunk 17]
```diff
@@ -82,51 +82,39 @@ def change
       t.index :active_storage_attachment_id
     end
 
-    create_range_partition :impressions, id: :uuid, default: -> { "gen_random_uuid()" }, partition_key: "displayed_at_date" do |t|
```
Comment (coderberry) [inline db/migrate/20181017152837_shotgun.rb]: Maybe ad note to README that you are re-consolidating the shotgun. Then we can remove the note once we stop doing that

File: app/models/concerns/campaigns/budgetable.rb [hunk 18]
```diff
@@ -0,0 +1,34 @@
+module Campaigns
+  module Budgetable
+    extend ActiveSupport::Concern
+
+    def total_consumed_budget
+      ecpm * total_impressions_per_mille
```
[the remark this answers is not in the document]
Comment (hopsoft) [reply app/models/concerns/campaigns/budgetable.rb]: Added note to circle back on this.

File: app/models/impression.rb (hunk 9 above)
Comment (hopsoft) [reply app/models/impression.rb]: Lets circle back to this. I think we need to add the column `clicked_at_date` and index that one.

File: db/migrate/20181017152837_shotgun.rb (hunk 17 above)
Comment (hopsoft) [reply db/migrate/20181017152837_shotgun.rb]: Decided not to do this

**Pull request merged**